Source/Mock.cs

```diff
@@ -838,44 +837,47 @@ private static void SetupAllProperties(Mock mock, Stack<Type> mockedTypesStack)
 			{
 				var expression = GetPropertyExpression(mockType, property);
 				var getter = property.GetGetMethod(true);
-				object value = SetupAllProperties_GetInitialValue(mock, mockedTypesStack, getter);
 
-				var mocked = value as IMocked;
-				if (mocked != null)
+				object value = null;
+				bool valueNotSet = true;
+
+				mock.Setups.Add(new PropertyGetterMethodCall(mock, expression, getter, () =>
 				{
-					SetupAllProperties(mocked.Mock, mockedTypesStack);
-				}
+					if (valueNotSet)
+					{
+						object initialValue;
+						try
+						{
+							initialValue = mock.GetDefaultValue(getter, useAlternateProvider: defaultValueProvider);
+						}
+						catch
+						{
+							// Since this method performs a batch operation, a single failure of the default value
+							// provider should not tear down the whole operation. The empty default value provider
+							// is a safe fallback because it does not throw.
+							initialValue = mock.GetDefaultValue(getter, useAlternateProvider: DefaultValueProvider.Empty);
+						}
+
+						if (initialValue is IMocked mocked)
+						{
+							SetupAllPropertiesPexProtected(mocked.Mock, defaultValueProvider);
+						}
 
-				mock.Setups.Add(new PropertyGetterMethodCall(mock, expression, getter, () => value));
+						value = initialValue;
+						valueNotSet = false;
+					}
+					return value;
+				}));
 
 				if (property.CanWrite)
 				{
-					mock.Setups.Add(new PropertySetterMethodCall(mock, expression, property.GetSetMethod(true), (newValue) => value = newValue));
+					mock.Setups.Add(new PropertySetterMethodCall(mock, expression, property.GetSetMethod(true), (newValue) =>
+					{
+						value = newValue;
+						valueNotSet = false;
+					}));
 				}
 			}
-
-			mockedTypesStack.Pop();
-		}
-
-		private static object SetupAllProperties_GetInitialValue(Mock mock, Stack<Type> mockedTypesStack, MethodInfo getter)
-		{
-			if (mockedTypesStack.Contains(getter.ReturnType))
-			{
-				// to deal with loops in the property graph
-				return mock.GetDefaultValue(getter, useAlternateProvider: DefaultValueProvider.Empty);
```

Comment:
We no longer need to guard against "loops in the property graph" since `SetupAllProperties` only sets up the properties of the current level in the object's hierarchy. The setting up of deeper levels in the hierarchy must be triggered by querying a property.